### PR TITLE
Add EntityRef/EntityMut based options to ReflectComponent

### PR DESCRIPTION
**Pull request by james7132**
# Objective
`ReflectComponent::get` and friends internally use `World::get` and similar APIs which spend time looking up the `EntityLocation` on each call. This can add significant overhead when fetching multiple components from the same entity, such as when animating multiple components via reflection.

## Solution
Extend the functions of `ReflectComponent` to fetch references to components using `EntityRef` and `EntityMut` instead of a World/Entity. This allows users to fetch a single `EnttiyRef`/`EntityMut` and query for any or all of it's components without refetching the entity's location for each one.

Ideally, a World-specific version of `ReflectComponent` would cache the `ComponentId` so we could cut out the `TypeId` -> `ComponentId` lookup as well. However this requires a `&mut World` on `ReflectComponent` initialization and would only be valid for that specific World. 

---

## Changelog
Added: `ReflectComponent::reflect_ref`
Added: `ReflectComponent::reflect_mut_ref`
Added: `ReflectComponent::reflect_unchecked_mut_ref`

### CODE REVIEW

File: crates/bevy_ecs/src/reflect.rs [hunk 1]
```diff
@@ -106,11 +112,22 @@ impl ReflectComponent {
     }
 
     /// Gets the value of this [`Component`] type from the entity as a reflected reference.
+    ///
+    /// If repeatedly fetching multiple components from the same entity, prefer using
+    /// [`ReflectComponent::reflect_ref`] where possible to avoid multiple entity location lookups.
     pub fn reflect<'a>(&self, world: &'a World, entity: Entity) -> Option<&'a dyn Reflect> {
         (self.0.reflect)(world, entity)
```
Comment (JoJoJet) [inline crates/bevy_ecs/src/reflect.rs:119]: Could this be changed to use the same function pointer as `reflect_ref`?

Comment (james7132) [reply crates/bevy_ecs/src/reflect.rs:119]: That would require returning a reference bound to the lifetime of the dropped EntityRef.

Comment (JoJoJet) [reply crates/bevy_ecs/src/reflect.rs:119]: Apparently that is only true for the mutable variant -- the read-only version is bound to the lifetime of the `World`, not the `EntityRef`. Not entirely necessary though, I suppose.

File: crates/bevy_ecs/src/reflect.rs [hunk 2]
```diff
@@ -221,6 +270,25 @@ impl<C: Component + Reflect + FromWorld> FromType<C> for ReflectComponent {
...
+            reflect_mut_ref: |entity_mut| {
+                entity_mut.get_mut::<C>().map(|c| Mut {
+                    value: c.value as &mut dyn Reflect,
+                    ticks: c.ticks,
+                })
+            },
+            reflect_mut_unchecked_ref: |entity_ref| {
+                let world = entity_ref.world();
+                // SAFETY: reflect_mut_ref is an unsafe function pointer used by `reflect_unchecked_mut_ref ` which promises to never
+                // produce aliasing mutable references, and reflect_mut_ref, which has mutable world access
```
Comment (JoJoJet) [inline crates/bevy_ecs/src/reflect.rs:282]: This safety comment looks incorrect?

**Comment by james7132** [comment]
I have not, though my gut feeling like it's going to be slower due to not having any of the compile time optimizations we've been adding to the generic path. The use of `*_by_id` functions also puts `unsafe` in the userspace code more readily, so I'm hesitant to suggest it's usage.

**Comment by james7132** [comment]
> This could be fixed by having wrappers at a safe boundary, like (TypeId, &TypeRegistry) -> &dyn Reflect

This would still lose the `EntityRef::get<T>` vs `EntityRef::get_by_id` difference where it uses the compile time constant to optimize out a branch. This would have a negative impact on "animate anything" which will be highly reliant on this API to be performant.

**Comment by james7132** [comment]
> Why did you keep the `reflect_mut` function pointer instead of implementing it in terms of the underlying `EntityMut` one?

The primary one is that the lifetime of a `EntityMut` and it's fetched borrows is held in terms of where the it's created from. Creating an `EntityMut` inside that function leaks the lifetime to the outside, so it doesn't compile.

**Comment by james7132** [comment]
Closing this as it's already been implemented by #7206.

**Pull request closed without merging**